internal/client/client_encrypted.go

```diff
@@ -0,0 +1,233 @@
+package client
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/tidwall/gjson"
+	"maunium.net/go/mautrix"
+	"maunium.net/go/mautrix/crypto"
+	"maunium.net/go/mautrix/event"
+	"maunium.net/go/mautrix/id"
+
+	"github.com/matrix-org/complement/internal/must"
+)
+
+type CSAPIEncrypted struct {
+	*CSAPI
+	mautrix    *mautrix.Client
+	olmMachine *crypto.OlmMachine
+	stateStore StateStore
+	encChan    chan EncryptedEventResponse
+}
+
+type EncryptedEventResponse struct {
+	Body  string
+	Error error
+}
+
+type clientOptions struct {
+	logger crypto.Logger
+}
+
+type ClientOption func(*clientOptions)
+
+// WithLogging enables a crypto.Logger using t.Log/t.Fatal etc.
+func WithLogging(t *testing.T, csapi *CSAPI) ClientOption {
+	return func(opts *clientOptions) {
+		opts.logger = testCryptoLogger{
+			t:      t,
+			userID: csapi.UserID,
+		}
+	}
+}
+
+// MustCreateEncryptedClient creates an "encrypted client" from a CSAPI and initialises
+// the OlmMachine, fails the test if the client or OlmMachine can't be created
+func MustCreateEncryptedClient(t *testing.T, csapi *CSAPI, opts ...ClientOption) *CSAPIEncrypted {
+	defaultOptions := &clientOptions{
+		logger: emptyCryptoLogger{},
+	}
+
+	for _, opt := range opts {
+		opt(defaultOptions)
+	}
+
+	cl, err := mautrix.NewClient(csapi.BaseURL, id.UserID(csapi.UserID), csapi.AccessToken)
+	if err != nil {
+		t.Fatalf("failed to create mautrix.Client: %s", err)
+	}
+	cl.DeviceID = id.DeviceID(csapi.DeviceID)
+
+	stateStore := StateStore{storer: mautrix.NewInMemoryStore()}
+
+	// Open SQL connection
+	db, err := sql.Open("sqlite3", ":memory:")
+	if err != nil {
+		t.Fatalf("failed to open sql.DB: %s", err)
+	}
+
+	accountID := csapi.UserID + "-" + csapi.DeviceID
+	sqlCryptoStore := crypto.NewSQLCryptoStore(db, "sqlite3", accountID, id.DeviceID(csapi.DeviceID), []byte(csapi.DeviceID+"pickle"), defaultOptions.logger)
+
+	// Create needed tables
+	if err = sqlCryptoStore.CreateTables(); err != nil {
+		t.Fatalf("unable to create crypto store tables: %s", err)
+	}
+
+	olmMachine := crypto.NewOlmMachine(cl, defaultOptions.logger, sqlCryptoStore, &stateStore)
+
+	encAPI := &CSAPIEncrypted{
+		CSAPI:      csapi,
+		olmMachine: olmMachine,
+		mautrix:    cl,
+		stateStore: stateStore,
+		encChan:    make(chan EncryptedEventResponse),
+	}
+
+	// Accept all incoming verification requests
+	/*olmMachine.AcceptVerificationFrom = func(_ string, otherDevice *crypto.DeviceIdentity, _ id.RoomID) (crypto.VerificationRequestResponse, crypto.VerificationHooks) {
+		return crypto.AcceptRequest, encAPI
+	}*/
+
+	syncer := cl.Syncer.(*mautrix.DefaultSyncer)
+	// Handle general events, e.g. to update the olmMachine and stateStore
+	syncer.OnSync(encAPI.syncCallback(t))
+
+	// Handle membership events
+	syncer.OnEventType(event.StateMember, func(_ mautrix.EventSource, evt *event.Event) {
+		encAPI.olmMachine.HandleMemberEvent(evt)
```

Comment:
We need to be treating this in the same way that we treat the federation server API wise. In other words, by default do _nothing fancy_. Expose optional _opt-in_ mechanics for "automatically keeping member lists up to date" and so on.